### ISSUE TITLE
Can no longer log in: Bad username or password 

### ISSUE THREAD

**Issue by boredazfcuk** [icloud-photos-downloader/icloud_photos_downloader#729]
## Overview
As of about 45mins ago, all four of my containers attempted a download, and all four failed to login.

After removing the keyring file and attempting to recreate it with `icloud --username my@email.com` the password is rejected. I've confirmed the e-mail and password combination is correct.

## Steps to Reproduce
[NOTE]: # ( Provide a simple set of steps to reproduce this bug. )
1. Start container

## Expected Behavior
Logs into icloud

## Actual Behavior
Fails to login. If password is saved to the keyring, this error is generated:

```
  File "/opt/icloudpd_latest/lib/python3.11/site-packages/pyicloud_ipd/base.py", line 220, in authenticate
    req = self.session.post(
          ^^^^^^^^^^^^^^^^^^
  File "/opt/icloudpd_latest/lib/python3.11/site-packages/requests/sessions.py", line 637, in post
    return self.request("POST", url, data=data, json=json, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/icloudpd_latest/lib/python3.11/site-packages/pyicloud_ipd/base.py", line 105, in request
    self._raise_error(code, reason)
  File "/opt/icloudpd_latest/lib/python3.11/site-packages/pyicloud_ipd/base.py", line 127, in _raise_error
    raise api_error
pyicloud_ipd.exceptions.PyiCloudAPIResponseError: Unknown reason
During handling of the above exception, another exception occurred:
Traceback (most recent call last):
  File "/opt/icloudpd_latest/bin/icloudpd", line 8, in <module>
    sys.exit(main())
             ^^^^^^
  File "/opt/icloudpd_latest/lib/python3.11/site-packages/click/core.py", line 1157, in __call__
    return self.main(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/icloudpd_latest/lib/python3.11/site-packages/click/core.py", line 1078, in main
    rv = self.invoke(ctx)
         ^^^^^^^^^^^^^^^^
  File "/opt/icloudpd_latest/lib/python3.11/site-packages/click/core.py", line 1434, in invoke
    return ctx.invoke(self.callback, **ctx.params)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/icloudpd_latest/lib/python3.11/site-packages/click/core.py", line 783, in invoke
    return __callback(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/icloudpd_latest/lib/python3.11/site-packages/icloudpd/base.py", line 317, in main
    core(
  File "/opt/icloudpd_latest/lib/python3.11/site-packages/icloudpd/base.py", line 744, in core
    icloud = authenticator(logger, domain)(
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/icloudpd_latest/lib/python3.11/site-packages/icloudpd/authentication.py", line 31, in authenticate_
    icloud = pyicloud_ipd.PyiCloudService(
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/icloudpd_latest/lib/python3.11/site-packages/pyicloud_ipd/base.py", line 204, in __init__
    self.authenticate()
  File "/opt/icloudpd_latest/lib/python3.11/site-packages/pyicloud_ipd/base.py", line 228, in authenticate
    raise PyiCloudFailedLoginException(msg, error)
pyicloud_ipd.exceptions.PyiCloudFailedLoginException: ('Invalid email/password combination.', PyiCloudAPIResponseError('Unknown reason'))
```

## Context
Seems that Apple may have changed something. Can replicate on Alpine Linux 3.18.3 icloudpd 1.16.2 and Alpine Linux 3.18.5 and icloudpd 1.16.3

**Comment by gunner007cd** [comment]
Same issue was well, failure to login, even after changing password to something simple.

**Comment by AndreyNikiforov** [comment]
I can replicate the issue on docker with 1.16.3 without keyring, so it is general issue and most likely related to some changes on Apple side as my 1.16.3 was running for a number of days before the issue started.

**Comment by AnthonySchneider-cnic** [comment]
Same for me using docker on Ubuntu. This was working 2 or 3 days ago. Tried with version 1.16.2 and 1.16.3.
A password reset did not help.

**Comment by OZidkani** [comment]
Same here

**Comment by seankerrigan** [comment]
also experiencing the same issue with 1.16.2 as well as after upgrading to 1.16.3

**Comment by jckesser** [comment]
Same here

**Comment by benstn** [comment]
Same issue - container has stopped downloading today

**Comment by oliplot** [comment]
Same issue here, stopped working yesterday Dec 6

**Comment by gxander85** [comment]
Following

**Comment by thiagokaplar** [comment]
Following

**Comment by petebocken** [comment]
Same here, following

**Comment by eastexxiao** [comment]
Same, following;

**Comment by staffankvisth** [comment]
I also get the same error on two different iCloud accounts. Suddenly stopped working 2 days ago. Both of my boredazfcuk/docker-icloudpd containers stopped and also manual execution of the script
_~/.local/bin/icloudpd -u [my@email.address](mailto:my@email.address) -p ****** -d . --folder-structure={:%Y/%m/%d} --delete-after-download_
fails with:
_2023-12-07 23:24:36 DEBUG Authenticating... 2023-12-07 23:24:36 ERROR Unknown reason Traceback (most recent call last): File "pyicloud_ipd/base.py", line 220, in authenticate File "requests/sessions.py", line 637, in post File "pyicloud_ipd/base.py", line 105, in request File "pyicloud_ipd/base.py", line 127, in _raise_error pyicloud_ipd.exceptions.PyiCloudAPIResponseError: Unknown reason

During handling of the above exception, another exception occurred:

Traceback (most recent call last): File "starters/icloudpd.py", line 5, in File "click/core.py", line 1157, in call File "click/core.py", line 1078, in main File "click/core.py", line 1434, in invoke File "click/core.py", line 783, in invoke File "icloudpd/base.py", line 317, in main File "icloudpd/base.py", line 744, in core File "icloudpd/authentication.py", line 31, in authenticate_ File "pyicloud_ipd/base.py", line 204, in init File "pyicloud_ipd/base.py", line 228, in authenticate pyicloud_ipd.exceptions.PyiCloudFailedLoginException: ('Invalid email/password combination.', PyiCloudAPIResponseError('Unknown reason')) [1357605] Failed to execute script 'icloudpd' due to unhandled exception!_

Logging in on the icloud.com website works as it should.
Does this mean there is no remedy for this problem yet? Has Apple changed something?

**Comment by syunlee66** [comment]
Same, following;

**Comment by Designator-Ol** [comment]
just following.

**Comment by MaPa80711** [comment]
Hi, maybe it will help to solve the problem. I have three icloud Accounts running and the issue is the same in every account.
 
icloudpd --directory /xxx/xxx/xxx/icloudpd_photo_backup --username xxxxxxxxxxxxx@icloud.com --password xxxxxxxx --log-level debug
2023-12-08 16:16:11 DEBUG    Authenticating...
2023-12-08 16:16:12 ERROR    Unknown reason
Traceback (most recent call last):
  File "pyicloud_ipd/base.py", line 220, in authenticate
  File "requests/sessions.py", line 637, in post
  File "pyicloud_ipd/base.py", line 105, in request
  File "pyicloud_ipd/base.py", line 127, in _raise_error
pyicloud_ipd.exceptions.PyiCloudAPIResponseError: Unknown reason

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "starters/icloudpd.py", line 5, in <module>
  File "click/core.py", line 1157, in __call__
  File "click/core.py", line 1078, in main
  File "click/core.py", line 1434, in invoke
  File "click/core.py", line 783, in invoke
  File "icloudpd/base.py", line 317, in main
  File "icloudpd/base.py", line 744, in core
  File "icloudpd/authentication.py", line 31, in authenticate_
  File "pyicloud_ipd/base.py", line 204, in __init__
  File "pyicloud_ipd/base.py", line 228, in authenticate
pyicloud_ipd.exceptions.PyiCloudFailedLoginException: ('Invalid email/password combination.', PyiCloudAPIResponseError('Unknown reason'))
[201833] Failed to execute script 'icloudpd' due to unhandled exception!

**Comment by AndreyNikiforov** [comment]
I see the following paths:

1. dig into current code/protocol to see if minor tweak will make it working
   - rationale: fastest and least invasive
   - concerns: low chances
   - chances: low; suspect Apple turned off old protocol that `icloudpd` uses (they support OAUTH on web for some time)
   - investments: low
2. use original pyicloud lib
   - rationale: IIUC pyicloud was updated to support OAUTH some time ago
   - concerns: may break some of our code
   - chances: mid-high; need to confirm that pyicloud works with current Apple services first
   - investments: mid; need to update tests and that has been a challenge in the prev attempts to bring latest pyicloud
3. use own code to support latest Apple auth
   - rationale: guaranteed success, opens door for other long-requested improvements
   - concerns: may force larger updates, leading to breaking changes, long alpha-beta cycle
   - chances: high
   - investments: high

Anybody interested in 1) and/or 2)? I plan to allocate some time in the next number of days to look into this issue and will invest in 1 & 3. IMO it is okay for multiple ppl to look at the same path as there are always edge cases and different approaches and may give different or complementing outcomes.

**Comment by robflate** [comment]
Total speculation but I wonder if this has anything to do with Apple's attempts to block iMessage on Android;

https://www.theverge.com/2023/12/8/23994089/apple-beeper-mini-android-blocked-imessage-app

**Comment by nscheer** [comment]
My guess is that it has something to do with the recent support for passkey logins.
At least for the iCloud website it asks for passkey and logging in via password takes one more click.

Such a thing would be unusual for an api, nevertheless a change that requires specification of the desired login method sounds very likely.

**Comment by cutzenfriend** [comment]
> My guess is that it has something to do with the recent support for passkey logins. At least for the iCloud website it asks for passkey and logging in via password takes one more click.
> 
> Such a thing would be unusual for an api, nevertheless a change that requires specification of the desired login method sounds very likely.

This sounds like the issue for me...

**Comment by charlesb87** [comment]
Same error here running on Windows x64 with 1.16.3 version

**Comment by dlang123** [comment]
Same issue for me as well, docker that was previously working fine has now failed authorization today as well, and I haven't been able to authorize one for 3 days.

**Comment by mattryles** [comment]
This looks related to beeper mini being blocked. 😔

**Comment by scaraebeus** [comment]
> I see the following paths:
> 
> 1. dig into current code/protocol to see if minor tweak will make it working
> 
> * rationale: fastest and least invasive
> * concerns: low chances
> * chances: low; suspect Apple turned off old protocol that `icloudpd` uses (they support OAUTH on web for some time)
> * investments: low
> 
> 2. use original pyicloud lib
> 
> * rationale: IIUC pyicloud was updated to support OAUTH some time ago
> * concerns: may break some of our code
> * chances: mid-high; need to confirm that pyicloud works with current Apple services first
> * investments: mid; need to update tests and that has been a challenge in the prev attempts to bring latest pyicloud
> 
> [ . . . ]
> 
> Is anybody interested in 1) and/or 2)? I plan to allocate some time in the next number of days to look into this issue and will invest in 1 & 3. IMO it is okay for multiple ppl to look at the same path as there are always edge cases and different approaches that may give different or complementing outcomes.

If it helps, I was able to verify the current 1.0.0 version of pyicloud works as is.  I was able to instantiate a PyiCloudService and list albums (pyicloud,photos.albums).

Potentially doing a combination of path 1) and 2) may be an option - possibly just focusing on merging over the key authentication pieces from the 1.0.0 pyicloud.

**Comment by cfurrow** [comment]
> If it helps, I was able to verify the current 1.0.0 version of pyicloud works as is. I was able to instantiate a PyiCloudService and list albums (pyicloud,photos.albums).

I was noticing that as well. `icloud_photos_downloader` uses some customized versions of `pyicloud` code to be compatible with China vs US icloud domains. It's possible those customizations are now causing some issues, and may not be updated with the latest code from `pyicloud`.

**Comment by wildenrou** [comment]
following - thanks

**Comment by Junod972** [comment]
Thanks a lot. Because, all this have a "family" impact. I need to have my dear photos and videos close to me on my NAS. If I can contribute one way or the other I will.
Thanks a looooooot. Merci 🤗.

**Comment by cfurrow** [comment]
There is hope for a fix, and would appreciate others testing if they have a chance: https://github.com/icloud-photos-downloader/icloud_photos_downloader/pull/733

Notes:
- many automated tests are broken, and need fixing
- there is currently no way to select .com vs .cn domains for icloud
- there is currently no way to select a photo library, so "All Photos" is used by default

I've had success running a docker container and pulling down all photos/movies from my iCloud photo library:

```sh
docker build . -t icloudpd_dev

docker run -it \
  --rm --name icloudpd \
  -v ~/icloudpd-download:/data \
  -e TZ=America/New_York \
  icloudpd_dev \
    icloudpd \
    --directory /data \
    --username my-email@example.com \
    --watch-with-interval 3600

2023-12-09 16:23:42 DEBUG    Authenticating...
iCloud Password:
2023-12-09 16:24:00 INFO     Two-factor authentication is required
Please enter two-factor authentication code: 123456

2023-12-09 16:24:07 WARNING  Failed to parse response with JSON mimetype
2023-12-09 16:24:10 INFO     Great, you're all set up. The script can now be run without user interaction until 2SA expires.
You can set up email notifications for when the two-step authentication expires.
(Use --help to view information about SMTP options.)
2023-12-09 16:24:13 DEBUG    Looking up all photos and videos from album All Photos...
2023-12-09 16:24:13 INFO     Downloading 81324 original photos and videos to /data ...
2023-12-09 16:24:16 DEBUG    Downloading /data/2023/12/09/IMG_1083.MOV...
2023-12-09 16:24:18 INFO     Downloaded /data/2023/12/09/IMG_1083.MOV
2023-12-09 16:24:18 DEBUG    Downloading /data/2023/12/09/IMG_1082.JPG...
```

**Comment by bnutzer** [comment]
Hi,

the fix almost works. I have not configured a domain, and the current `auth_fix` branch fails with this error:

```
[...]
  File "/home/xxxx/.local/lib/python3.11/site-packages/pyicloud_ipd/base.py", line 393, in _authenticate_with_token
    domain_to_use = req.get('domainToUse')
                    ^^^^^^^
AttributeError: 'Response' object has no attribute 'get'
```

I have not fully understood what the `req.get` statement is supposed to do and expect there should be some kind of key check first.

My temporary fix is
```
        domain_to_use = None
        try:
            domain_to_use = req.get('domainToUse')
        except AttributeError as error:
            print("domainToUse attribute not found.")
```

With this, I am able to download my icloud photos again 🎉

**Comment by scaraebeus** [comment]
> Hi,
> 
> the fix almost works. I have not configured a domain, and the current `auth_fix` branch fails with this error:
> 
> ```
> [...]
>   File "/home/xxxx/.local/lib/python3.11/site-packages/pyicloud_ipd/base.py", line 393, in _authenticate_with_token
>     domain_to_use = req.get('domainToUse')
>                     ^^^^^^^
> AttributeError: 'Response' object has no attribute 'get'
> ```
> 
> I have not fully understood what the `req.get` statement is supposed to do and expect there should be some kind of key check first.
> 
> [ . . . ]

I've fixed this in the latest commit.  The get function will handle missing keys gracefully, unfortunately, it was being called on an object that doesn't have the method - so even if the key existed, it would still fail with the `AttributeError`

`req.json()` does have the `get` method (as it returns a dictionary), and that is set to `self.data` just above, so calling `self.data.get("domainToUse")` should return `None` if it doesn't exist in the response.

**Comment by michaelmolino** [comment]
@scaraebeus' fix worked for me.  Built a local docker image and successfully grabbed my photos from two accounts!  🎉

**Comment by cfurrow** [comment]
@Braincoke 
You need to supply your own email to the `docker run` command:

```sh
docker run -it \
  --rm \
  -v $(pwd)/mynas/Photos/iCloudBackup:/data \
  -e TZ=America/New_York \
  icloudpd icloudpd --directory /data --username  myemail@mail.com                         
```

You can see that in the command above, I pass a dummy email, `--username  myemail@mail.com`
You would want to replace that with your actual apple id email.

**Comment by Braincoke** [comment]
Thanks @cfurrow, I did pass my real apple ID to the command.

**Comment by scaraebeus** [comment]
@Braincoke - I took a quick look and the icloudpd auth flow should prompt for a password if it's not supplied in the command line `--password your_password` and also not found in the keyring.

With the changes in the `pyicloud` exception names, this exception isn't being caught by icloudpd to prompt for the password.  It was previously looking for `pyicloud_ipd.exceptions.NoStoredPasswordAvailable` and should now be looking for `pyicloud_ipd.exceptions.PyiCloudNoStoredPasswordAvailableException`.

I'll review the `pyicloud` exceptions referenced throughout `icloudpd` and update.

**Comment by vw-kombi** [comment]
I am on unraid and hence cant test any of this stuff being reported, but looking forward to testing it once an update goes to the unraid app store.

**Comment by winteriscariot** [comment]
The docker solution does not seem to work for me: I'm still getting invalid email/password combo when attempting this via a build docker container:

```
  560  git clone https://github.com/icloud-photos-downloader/icloud_photos_downloader.git
  561  cd icloud_photos_downloader/
  562  ls
  563  docker build . -t icloudpd_dev
  564  history
  565  docker run -it --rm --name icloudpd -v ~/Pictures/Photos:/data -e TZ=America/Vancouver icloudpd_dev icloudpd --directory /data --username my@email.com --watch-with-interval 1800
2023-12-12 07:35:41 DEBUG    Authenticating...
iCloud Password: 
2023-12-12 07:35:44 ERROR    Unknown reason
Traceback (most recent call last):
  File "pyicloud_ipd/base.py", line 220, in authenticate
  File "requests/sessions.py", line 637, in post
  File "pyicloud_ipd/base.py", line 105, in request
  File "pyicloud_ipd/base.py", line 127, in _raise_error
pyicloud_ipd.exceptions.PyiCloudAPIResponseError: Unknown reason

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "starters/icloudpd_ex.py", line 110, in <module>
  File "starters/icloudpd_ex.py", line 106, in main
  File "click/core.py", line 1157, in __call__
  File "click/core.py", line 1078, in main
  File "click/core.py", line 1688, in invoke
  File "click/core.py", line 1434, in invoke
  File "click/core.py", line 783, in invoke
  File "icloudpd/base.py", line 317, in main
  File "icloudpd/base.py", line 744, in core
  File "icloudpd/authentication.py", line 31, in authenticate_
  File "pyicloud_ipd/base.py", line 204, in __init__
  File "pyicloud_ipd/base.py", line 228, in authenticate
pyicloud_ipd.exceptions.PyiCloudFailedLoginException: ('Invalid email/password combination.', PyiCloudAPIResponseError('Unknown reason'))
[8] Failed to execute script 'icloudpd_ex' due to unhandled exception!
~/.src/icloud_photos_downloader $> 
```

**Comment by scaraebeus** [comment]
> The docker solution does not seem to work for me: I'm still getting invalid email/password combo when attempting this via a build docker container:

The master branch has yet to be updated.  One proposed solution is available in PR #734.

**Comment by lcmartos** [comment]
@winteriscariot I think you're not getting the repo with the fix. The PR that is waiting to merge is https://github.com/scaraebeus/icloud_photos_downloader.git right now.

Also, this don't work for me. I've got the same error:


```
git clone https://github.com/scaraebeus/icloud_photos_downloader.git

docker build . -t icloudpd_scarabeus --no-cache

docker run -it    --rm --name icloudpd    -v /tmp/icloudpd-download:/data    -e TZ=America/New_York    icloudpd_scarabeus      icloudpd      --directory /data      --username USER@DOMAIN.COM      --watch-with-interval 3600 --password "PASSWORD"
2023-12-12 11:24:26 DEBUG    Authenticating...
2023-12-12 11:24:27 ERROR    Unknown reason
Traceback (most recent call last):
  File "pyicloud_ipd/base.py", line 220, in authenticate
  File "requests/sessions.py", line 637, in post
  File "pyicloud_ipd/base.py", line 105, in request
  File "pyicloud_ipd/base.py", line 127, in _raise_error
pyicloud_ipd.exceptions.PyiCloudAPIResponseError: Unknown reason

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "starters/icloudpd_ex.py", line 110, in <module>
  File "starters/icloudpd_ex.py", line 106, in main
  File "click/core.py", line 1157, in __call__
  File "click/core.py", line 1078, in main
  File "click/core.py", line 1688, in invoke
  File "click/core.py", line 1434, in invoke
  File "click/core.py", line 783, in invoke
  File "icloudpd/base.py", line 317, in main
  File "icloudpd/base.py", line 744, in core
  File "icloudpd/authentication.py", line 31, in authenticate_
  File "pyicloud_ipd/base.py", line 204, in __init__
  File "pyicloud_ipd/base.py", line 228, in authenticate
pyicloud_ipd.exceptions.PyiCloudFailedLoginException: ('Invalid email/password combination.', PyiCloudAPIResponseError('Unknown reason'))
[7] Failed to execute script 'icloudpd_ex' due to unhandled exception!
```

**Comment by michaelmolino** [comment]
@lcmartos `git checkout auth_fix`

**Comment by lcmartos** [comment]
@michaelmolino   Working! Thank you! I suppose it's just a matter of time before they try and accept the patch.

**Comment by fasihi01** [comment]
I got it working too - following steps did the trick for me (checking out the head "auth_fix" is the trick really):
wget https://github.com/scaraebeus/icloud_photos_downloader/archive/refs/heads/auth_fix.zip
mv auth_fix auth_fix.zip
unzip auth_fix
docker build . -t icloudpd_dev
docker run -it --rm --name icloudpd -v /mnt/fotos:/data -e TZ=America/Vancouver icloudpd_dev icloudpd --directory /data --username mail@mail.com --watch-with-interval 1800 --password XXX

**Comment by espro1** [comment]
> I got it working too - following steps did the trick for me (checking out the head "auth_fix" is the trick really): wget https://github.com/scaraebeus/icloud_photos_downloader/archive/refs/heads/auth_fix.zip mv auth_fix auth_fix.zip unzip auth_fix docker build . -t icloudpd_dev docker run -it --rm --name icloudpd -v /mnt/fotos:/data -e TZ=America/Vancouver icloudpd_dev icloudpd --directory /data --username [mail@mail.com](mailto:mail@mail.com) --watch-with-interval 1800 --password XXX

I followed these steps and get the error  "Missing apple_id field":

`2023-12-12 16:03:27 DEBUG    Authenticating...
2023-12-12 16:03:28 ERROR    Missing apple_id field
Traceback (most recent call last):
  File "pyicloud_ipd/base.py", line 365, in _authenticate_with_token
  File "requests/sessions.py", line 637, in post
  File "pyicloud_ipd/base.py", line 156, in request
  File "pyicloud_ipd/base.py", line 185, in _raise_error
pyicloud_ipd.exceptions.PyiCloudAPIResponseException: Missing apple_id field

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "starters/icloudpd_ex.py", line 110, in <module>
  File "starters/icloudpd_ex.py", line 106, in main
  File "click/core.py", line 1157, in __call__
  File "click/core.py", line 1078, in main
  File "click/core.py", line 1688, in invoke
  File "click/core.py", line 1434, in invoke
  File "click/core.py", line 783, in invoke
  File "icloudpd/base.py", line 317, in main
  File "icloudpd/base.py", line 744, in core
  File "icloudpd/authentication.py", line 31, in authenticate_
  File "pyicloud_ipd/base.py", line 283, in __init__
  File "pyicloud_ipd/base.py", line 345, in authenticate
  File "pyicloud_ipd/base.py", line 371, in _authenticate_with_token
pyicloud_ipd.exceptions.PyiCloudFailedLoginException: ('Invalid authentication token.', PyiCloudAPIResponseException('Missing apple_id field'))
[7] Failed to execute script 'icloudpd_ex' due to unhandled exception!`

I'm sure this is basic user error on my part but I thought I would share in case it isn't.

**Comment by scaraebeus** [comment]
> I followed these steps and get the error `Missing apple_id field`:
> 
> [ . . . ]
> 
> I'm sure this is basic user error on my part but I thought I would share in case it isn't.
> 
> I also have to pass the password as an option. If I don't, it complains that it isn't in the keyfile.

I'm not sure all of the ways this error can manifest - so this may not apply, however, in some of my testing I was able to get this error if passing an invalid username and/or incorrect password.  May want to double check there are no typos there.

**Comment by wildenrou** [comment]
> I got it working too - following steps did the trick for me (checking out the head "auth_fix" is the trick really): wget https://github.com/scaraebeus/icloud_photos_downloader/archive/refs/heads/auth_fix.zip mv auth_fix auth_fix.zip unzip auth_fix docker build . -t icloudpd_dev docker run -it --rm --name icloudpd -v /mnt/fotos:/data -e TZ=America/Vancouver icloudpd_dev icloudpd --directory /data --username [mail@mail.com](mailto:mail@mail.com) --watch-with-interval 1800 --password XXX

that worked for me and photos are syncing again. Thanks so much everyon for the great and quick team work on this one. Cheers.

**Comment by espro1** [comment]
> > I followed these steps and get the error `Missing apple_id field`:
> > [ . . . ]
> > I'm sure this is basic user error on my part but I thought I would share in case it isn't.
> > I also have to pass the password as an option. If I don't, it complains that it isn't in the keyfile.
> 
> I'm not sure all of the ways this error can manifest - so this may not apply, however, in some of my testing I was able to get this error if passing an invalid username and/or incorrect password. May want to double check there are no typos there.

Thanks for the tip. Just for the record, I've double- and triple-checked and I don't have any typos in username or password.

**Comment by vw-kombi** [comment]
Thanks for all the work on this everyone.
Seems like there are fixes in place for some.
I can test too it it gets to the unraid community app store.

**Comment by espro1** [comment]
I think I found the issue I was having. My password has a special character and if I escape it it works. I should note that I didn't have to do this in the icloudpd version I was running until a few days ago.

Thanks everyone :)

**Comment by thedinz** [comment]
> Thanks for all the work on this everyone.
> Seems like there are fixes in place for some.
> I can test too it it gets to the unraid community app store.

Yeah I use unraid and I have no idea how to apply the fix lol

**Comment by amdydesign** [comment]
Any solution yet?

**Comment by brettscott** [comment]
@amdydesign as above, `git checkout auth_fix`. Or wait until branch is merged into master.

**Comment by amdydesign** [comment]
> I got it working too - following steps did the trick for me (checking out the head "auth_fix" is the trick really): wget https://github.com/scaraebeus/icloud_photos_downloader/archive/refs/heads/auth_fix.zip mv auth_fix auth_fix.zip unzip auth_fix docker build . -t icloudpd_dev docker run -it --rm --name icloudpd -v /mnt/fotos:/data -e TZ=America/Vancouver icloudpd_dev icloudpd --directory /data --username [mail@mail.com](mailto:mail@mail.com) --watch-with-interval 1800 --password XXX

how to with this command add --delete-after-download, because its not working

**Comment by HeikoGr** [comment]
it's multiple commands. your "delete after download" belongs to icloudpd and belongs to the last command:

`wget https://github.com/scaraebeus/icloud_photos_downloader/archive/refs/heads/auth_fix.zip`

`mv auth_fix auth_fix.zip`

`unzip auth_fix`

`docker build . -t icloudpd_dev`

`docker run -it --rm --name icloudpd -v /mnt/fotos:/data -e TZ=America/Vancouver icloudpd_dev icloudpd --directory /data --username mail@mail.com --watch-with-interval 1800 --password XXX --delete-after-download`

**Comment by thedinz** [comment]
I really fine this just above my head, maybe I can do It but I dont want to mess things up, any idea when the unraid image would be ready, and I dont mean an exact time im justing wondering if anyone is working to update that part or is this manual fix is it, this is what we have either use the fix or dont?

**Comment by amdydesign** [comment]
> it's multiple commands. your "delete after download" belongs to icloudpd and belongs to the last command:
> 
> `wget https://github.com/scaraebeus/icloud_photos_downloader/archive/refs/heads/auth_fix.zip`
> 
> `mv auth_fix auth_fix.zip`
> 
> `unzip auth_fix`
> 
> `docker build . -t icloudpd_dev`
> 
> `docker run -it --rm --name icloudpd -v /mnt/fotos:/data -e TZ=America/Vancouver icloudpd_dev icloudpd --directory /data --username mail@mail.com --watch-with-interval 1800 --password XXX --delete-after-download`

its downloading photos but didnt delete them after download

**Comment by dohgren78** [comment]
> it's multiple commands. your "delete after download" belongs to icloudpd and belongs to the last command:
> 
> `wget https://github.com/scaraebeus/icloud_photos_downloader/archive/refs/heads/auth_fix.zip`
> 
> `mv auth_fix auth_fix.zip`
> 
> `unzip auth_fix`
> 
> `docker build . -t icloudpd_dev`
> 
> `docker run -it --rm --name icloudpd -v /mnt/fotos:/data -e TZ=America/Vancouver icloudpd_dev icloudpd --directory /data --username mail@mail.com --watch-with-interval 1800 --password XXX --delete-after-download`

Works with above for me on Linux and Docker. Great work regarding the workaround! All kudos!
But I can't figure out how to download shared libraries with this - is it possible?

**Comment by NGC3982** [comment]
> I got it working too - following steps did the trick for me (checking out the head "auth_fix" is the trick really): wget https://github.com/scaraebeus/icloud_photos_downloader/archive/refs/heads/auth_fix.zip mv auth_fix auth_fix.zip unzip auth_fix docker build . -t icloudpd_dev docker run -it --rm --name icloudpd -v /mnt/fotos:/data -e TZ=America/Vancouver icloudpd_dev icloudpd --directory /data --username [mail@mail.com](mailto:mail@mail.com) --watch-with-interval 1800 --password XXX

Can you elaborate on what exactly to do here? I can't move a file i don't have into a zip file, and so on..

**Comment by OrZidkani** [comment]
@NGC3982 
Here you go: https://github.com/OrZidkani/icloudpd_auth_fix

Make it easier to run

**Comment by lost-while-translating** [comment]
Is there a update for the PIP (windows) Version? I don't use docker.
Or do I have to wait for next friday?

Thanks for all the efforts to fix it.

Frank

**Comment by Eagle248** [comment]
Does anyone know how to apply this fix for docker running in synology nas?

**Comment by scaraebeus** [comment]
> Is there a update for the PIP (windows) Version? I don't use docker.
> Or do I have to wait for next friday?

Not sure on timing for the official update.  Unfortunately for my fix to be pulled in, there are a few things that need to be in place on my branch for it to be accepted.  I've been working through updating the various tests and code coverage and in doing so I am finding some other parts of the `pyicloud_ipd/base.py` auth flow that are no longer working correctly due to (possibly) recent changes in how the responses are coming through.

I'm working through it but it will take some time - especially with the holidays coming up.

The only option I'm aware of at this time is to consume the auth_fix branch directly if your environment and setup allows you - with the risk that it has yet to be fully tested and vetted.

**Comment by AndreyNikiforov** [comment]
@scaraebeus thanks a lot for your hard work and dedication. It is well appreciated by many.

**Comment by revpdev** [comment]
@scaraebeus I join previous comment to thank you for your great contribution here, I patched the main branch with #734, built and installed it with pip and got it working on both raspbian/debian bullseye (11) and bookworm (12) . I am also using edits as described in issue #249 so that I can download both edited media (the ones we see on the phone and in iCloud) and original media in a regular cron backup script to my pi-based NAS, which makes this script critical to me. thanks again and congrats for keeping it working !

**Comment by kohldavido** [comment]
> > Is there a update for the PIP (windows) Version? I don't use docker.
> > Or do I have to wait for next friday?
> 
> Not sure on timing for the official update. Unfortunately for my fix to be pulled in, there are a few things that need to be in place on my branch for it to be accepted. I've been working through updating the various tests and code coverage and in doing so I am finding some other parts of the `pyicloud_ipd/base.py` auth flow that are no longer working correctly due to (possibly) recent changes in how the responses are coming through.
> 
> I'm working through it but it will take some time - especially with the holidays coming up.
> 
> The only option I'm aware of at this time is to consume the auth_fix branch directly if your environment and setup allows you - with the risk that it has yet to be fully tested and vetted.

Works on Debian 12 with docker for me until some period of time where the Apple servers either: 
 hang up, as icloudpd appears to idle in the middle of downloading
 attempt negotiate 2fa and send me an authentication code while icloudpd is in the middle of downloading
In both cases there isn't any debug or info logging printed indicating why icloudpd paused, but my network traffic goes to near zero, so I'm guessing the sessions is closed by apple.

Additionally, I'm attempting to download over 180k photos, so that might be why apple is hanging up on me periodically. It doesn't explain the 2fa prompt while icloudpd is running and actively downloading though.

The expected behavior where re-running icloupd picks up where it left off results in diminishing returns:
1st session, 4k downloads. 
2nd session: 4k skipped and 2k additional downloads. 
3rd session: 6k skipped and 500 additional completed downloads.
And so on until no additional downloads complete at all. This might be from pyicloud json handling?

**Comment by yrahul3910** [comment]
Works! For those using the above solution, a minor change is necessary (I'm on macOS): in the Shell script, change `/mnt/iCloud` to `$(pwd)`. Then, in Docker Desktop, add the current directory under Settings --> Resources --> File sharing.

**Comment by tanookiben** [comment]
Is saving to the keychain not working for anyone else with the auth_fix branch?

```
Save password in keyring?  [y/N]: y
Two-step authentication required. Your trusted devices are:
  0: SMS to ********##
Which device would you like to use? [0]:
```

Previously I was able to see an option to use a 2FA code instead of SMS

**Comment by yrahul3910** [comment]
> Is saving to the keychain not working for anyone else with the auth_fix branch?
> 
> ```
> Save password in keyring?  [y/N]: y
> Two-step authentication required. Your trusted devices are:
>   0: SMS to ********##
> Which device would you like to use? [0]:
> ```
> 
> Previously I was able to see an option to use a 2FA code instead of SMS

Steps to reproduce? With the branch I just got a 2FA code on my devices like normal, and it prompted me in the terminal. I don't remember it giving me an option between SMS or 2FA code, though.

**Comment by tanookiben** [comment]
> > Is saving to the keychain not working for anyone else with the auth_fix branch?
> > ```
> > Save password in keyring?  [y/N]: y
> > Two-step authentication required. Your trusted devices are:
> >   0: SMS to ********##
> > Which device would you like to use? [0]:
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Previously I was able to see an option to use a 2FA code instead of SMS
> 
> Steps to reproduce? With the branch I just got a 2FA code on my devices like normal, and it prompted me in the terminal. I don't remember it giving me an option between SMS or 2FA code, though.

With `icloudpd --username "USERNAME" --password "PASSWORD"` it asks me for a 2FA right away

When I try to save in the keyring it only offers SMS instead of 2FA:

```
docker run --rm -it --name icloud \
  -v /tmp/icloudpd/data:/data \
  -v /tmp/icloudpd/cookies:/cookies \
  -e "TZ=America/Los_Angeles" \
  icloudpd:auth_fix \
  icloud --username USERNAME
Enter iCloud password for USERNAME: 
Save password in keyring?  [y/N]: y
Two-step authentication required. Your trusted devices are:
  0: SMS to ********##
Which device would you like to use? [0]
```

**Comment by scaraebeus** [comment]
It appears the command line flow for the base `pyicloud_ipd` service (which you are invoking with `icloud --username USERNAME`) was never updated in the `pyicloud - 1.0.0` release to use the new 2FA feature, so it prompts as if it is expecting 2SA (choosing a device to send the code to first).

It also appears, from what I can tell, the `icloudpd` flow doesn't offer to save in the keyring directly.

Based on the above, as you've saved the password in the keyring now, try using the `icloudpd` directly without the `--password` option and it should retrieve from the keyring - and ask for the 2FA without prompting for a device selection.

The option to offer saving in the keyring could be moved into the `icloudpd` flow if needed.  Not sure if there is already an issue for that or not.

**Comment by boredazfcuk** [comment]
Previous behaviour was that saving to the keychain required SMS MFA, it did not have an option for Apple's built in MFA.

Performing a download would then trigger a second MFA prompt, which had two options for MFA, SMS and Apple.

I noticed this behaviour when attempting to relocate the MFA cookie in my container from $HOME/.local to /config. Took me a lot of deletes/recreates to get it working reliably.

**Comment by scaraebeus** [comment]
> Previous behaviour was that saving to the keychain required SMS MFA, it did not have an option for Apple's built in MFA.
> 
> Performing a download would then trigger a second MFA prompt, which had two options for MFA; SMS and Apple.
> 
> I noticed this behaviour when attempting to relocate the MFA cookie in my container from $HOME/.local to /config. Took me a lot of deletes/recreates to get it working reliably.
> 
> Edit: I think it's also worth mentioning that I actually had two SMS numbers I could use for three of my containers. I have added my phone number as a trusted number to the three accounts of my family members. This allows me to re-authenticate their containers without me needing access to their devices. Very useful feature.

The current `pyicloud - 1.0.0` 2SA/2FA implementation doesn't appear to take this into account.  The command line implementation for `pyicloud` doesn't even test for 2FA (hence why it drops straight into asking which device to use).  The underlying hooks in the `pyicloud` 2FA implementation also limit the flexibility for the above use case.

On top of this, there are two separate endpoints used to validate codes depending on whether the 2SA or 2FA flow is triggered.  It's possible the endpoints don't really matter and the same endpoint can be used regardless - I'll have to test.

Proposed solution:

- Leave the default behavior as is and add a switch to list devices to alternatively send a code to
  - When initiating `icloudpd` if you have a default iDevice and the account is enabled for HSA2 (I believe all accounts are now enabled for HSA2 at this point), prompt for 2FA code directly as it should be auto sent to the device
    - This would likely cover a majority of use cases
    - There is an edge case where an account may not have an iDevice associated with it - in this case, there is likely still an SMS number to send to (potentially Android instead of iDevice).  This is already covered in the default behavior and automatically triggers the list device flow.
  - Adding a 'list devices' switch (`icloudpd --list-devices`) could trigger the flow to collect and show the available devices associated with the account, allowing the user to select a different device to send a code to (and displaying the option to bypass this and enter the code if you decide to go with the code auto sent to the default device)
    - With the above switch, a code would still likely be auto sent to an iDevice - choosing a different device (or SMS) to send a code to would need to be tested to verify which validation endpoint accepts the second code (manual request) - (or if both do)

**Comment by tanookiben** [comment]
> Based on the above, as you've saved the password in the keyring now, try using the `icloudpd` directly without the `--password` option and it should retrieve from the keyring - and ask for the 2FA without prompting for a device selection.

I think in theory this should work but the code I get via SMS can never be verified successfully:

```
Enter iCloud password for USERNAME: 
Save password in keyring?  [y/N]: y
Two-step authentication required. Your trusted devices are:
  0: SMS to ********##
Which device would you like to use? [0]: 0
Please enter validation code: ######
Failed to verify verification code
```

**Comment by scaraebeus** [comment]
I retract my statement about the `pyicloud - 1.0.0` not taking the 2SA/2FA into account, it does, the minimal fix in the auth_fix branch does not include the updated cmdline.py to expose this in the `pyicloud_ipd` direct invoking.

@tanookiben - when you attempt to verify the SMS code, have you also already received an auto-generated code to one of your iDevices?

If so, what happens if you enter that code at the prompt instead of the one sent through SMS?

This would at give me a bit of a clue as to where there is something end point related and/or if it's still expecting the first code.

**Comment by boredazfcuk** [comment]
I'm not near my laptop to test, but IIRC, mine crashed at the point where it attempted to list the phone numbers.

**Comment by scaraebeus** [comment]
Update: All test are now passing for the `auth_fix` branch with 99% code coverage (matching the master branch).  Hopefully this will allow for the branch to be pulled into the master.

The 2SA/2FA cases noted above will still need to be looked into but I'm hopeful these can be treated as separate issues and allow the core fix to be merged for those that are waiting on this due to various package distribution requirements.

Happy Holidays!

**Comment by AndreyNikiforov** [comment]
1.17.0 released and has the fix

**Comment by boredazfcuk** [comment]
I've built a new container with iCloud 1.17.0, but the behaviour is not the same. I am no longer presented with the option to perform SMS based multifactor authentication. I also receive a warning about it not being able to parse JSON, which I've not seen before but I'm not too worried about, as the application downloads the photos regardless:
```
2023-12-20 12:08:01 DEBUG    Configure password
2023-12-20 12:08:01 DEBUG    Adding password to keyring file: /config/python_keyring/keyring_pass.cfg
2023-12-20 12:08:01 DEBUG    Switched to icloudpd: 1.17.0
Enter iCloud password for my@email.com:
Save password in keyring? [y/N]: y

Two-step authentication required.
Please enter validation code
(string) --> 137446

2023-12-20 12:08:46 INFO     Starting container initialisation
2023-12-20 12:08:46 DEBUG    Generate MFA cookie using password stored in keyring file
2023-12-20 12:08:47 DEBUG    Switched to icloudpd: 1.17.0
2023-12-20 12:08:47 ERROR    Authentication required for Account. (421)
Please enter two-factor authentication code: 119033
2023-12-20 12:09:20 WARNING  Failed to parse response with JSON mimetype
```

Is SMS based multifactor authentication no longer possible?

**Comment by scaraebeus** [comment]
> Is SMS based multifactor authentication no longer possible?

With the new method used in `pyicloud - 1.0.0` it splits the flow between 2SA (using SMS) and 2FA (Auto-sent code) based on whether the iCloud account itself is set for HSA version 2 (it appears all accounts may now be set for this).

They way the base `pyicloud` does this check, assuming all accounts are now set as HSA version 2, there is actually no way the 2SA flow would ever be called.  I uncovered this awhile ago doing some other testing with an account that doesn't have an iDevice (using the free iCloud service as standalone).  In this scenario, I identified a flag noting if the account has a qualifying device.  Adding this in to the 2FA check (requiring this to be true) now allows the flow to fall back on the 2SA - but only if there isn't a qualifying device on the account.

I've started working on an enhancement to allow selecting an SMS for 2sa validation.  A 2fa code will still be sent to the iDevice automatically, but you should be able to choose an SMS, get a code there, and validate with that instead.

If you want to check that out, see the `2fa_enhancements` branch under `scaraebeus/icloud_photos_downloader` - keep in mind this is experimental though.  In that branch, you can use the `--list-devices` option to present the list of SMS devices and choose to send a code there (or choose to enter the auto sent 2FA code).

It may be worth starting a separate issue for this to be tracked outside of this thread.

**Comment by kmcphillips** [comment]
I depend on this tool and have been following this issue. Thanks so much for all the hard work and getting this fixed so quickly. You are appreciated! :pray:

**Comment by tanookiben** [comment]
> I've built a new container with iCloud 1.17.0, but the behaviour is not the same. I am no longer presented with the option to perform SMS based multifactor authentication. I also receive a warning about it not being able to parse JSON, which I've not seen before but I'm not too worried about, as the application downloads the photos regardless:
> 
> ```
> 2023-12-20 12:08:01 DEBUG    Configure password
> 2023-12-20 12:08:01 DEBUG    Adding password to keyring file: /config/python_keyring/keyring_pass.cfg
> 2023-12-20 12:08:01 DEBUG    Switched to icloudpd: 1.17.0
> Enter iCloud password for my@email.com:
> Save password in keyring? [y/N]: y
> 
> Two-step authentication required.
> Please enter validation code
> (string) --> 137446
> 
> 2023-12-20 12:08:46 INFO     Starting container initialisation
> 2023-12-20 12:08:46 DEBUG    Generate MFA cookie using password stored in keyring file
> 2023-12-20 12:08:47 DEBUG    Switched to icloudpd: 1.17.0
> 2023-12-20 12:08:47 ERROR    Authentication required for Account. (421)
> Please enter two-factor authentication code: 119033
> 2023-12-20 12:09:20 WARNING  Failed to parse response with JSON mimetype
> ```
> 
> Is SMS based multifactor authentication no longer possible?

Can you share the command you used here? I'm able to save credentials to the keyring with `icloud --username "USERNAME"` but `icloudpd --username "USERNAME"` without a `--password` flag always prompts me for a password

**Comment by dustinziska** [comment]
Just checking in on this issue, tried the lastest release today and it seems like it is still broken. Was https://github.com/icloud-photos-downloader/icloud_photos_downloader/pull/734, supposed to resolve this issue?

**Comment by AndreyNikiforov** [comment]
> Just checking in on this issue, tried the lastest release today and it seems like it is still broken

If you see the same issue then post here all details. Otherwise open new issue.
